File: app/src/main/java/com/krystianwsul/checkme/gui/friends/UserListFragment.java
package com.krystianwsul.checkme.gui.friends;


import android.graphics.Color;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.google.common.collect.Sets;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.gui.AbstractFragment;
import com.krystianwsul.checkme.gui.FabUser;
import com.krystianwsul.checkme.gui.SelectionCallback;
import com.krystianwsul.checkme.gui.tasks.FriendPickerFragment;
import com.krystianwsul.checkme.loaders.ShowProjectLoader;
import com.krystianwsul.checkme.persistencemodel.SaveService;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class UserListFragment extends AbstractFragment implements FabUser {
    private static final String PROJECT_ID_KEY = "projectId";

    private static final String SAVE_STATE_KEY = "saveState";

    private static final String FRIEND_PICKER_TAG = "friendPicker";

    private ProgressBar mFriendListProgress;
    private RecyclerView mFriendListRecycler;
    private TextView mEmptyText;

    @Nullable
    private String mProjectId;

    private FriendListAdapter mFriendListAdapter;

    @Nullable
    private ShowProjectLoader.Data mData;

    @Nullable
    private SaveState mSaveState;

    private final SelectionCallback mSelectionCallback = new SelectionCallback() {
        @Override
        protected void unselect() {
            mFriendListAdapter.unselect();
        }

        @Override
        protected void onMenuClick(MenuItem menuItem) {
            ArrayList<String> selectedUserDataEmails = mFriendListAdapter.getSelected();
            Assert.assertTrue(selectedUserDataEmails != null);
            Assert.assertTrue(!selectedUserDataEmails.isEmpty());

            switch (menuItem.getItemId()) {
                case R.id.action_custom_times_delete:
                    mFriendListAdapter.removeSelected();

                    break;
                default:
                    throw new UnsupportedOperationException();
            }
        }

        @Override
        protected void onFirstAdded() {
            ((AppCompatActivity) getActivity()).startSupportActionMode(this);

            mActionMode.getMenuInflater().inflate(R.menu.menu_custom_times, mActionMode.getMenu());

            updateFabVisibility();
        }

        @Override
        protected void onSecondAdded() {

        }

        @Override
        protected void onOtherAdded() {

        }

        @Override
        protected void onLastRemoved() {
            updateFabVisibility();
        }

        @Override
        protected void onSecondToLastRemoved() {

        }

        @Override
        protected void onOtherRemoved() {

        }
    };

    @Nullable
    private FloatingActionButton mFriendListFab;

    @NonNull
    public static UserListFragment newInstance() {
        return new UserListFragment();
    }

    public UserListFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        if (args != null) {
            Assert.assertTrue(args.containsKey(PROJECT_ID_KEY));

            mProjectId = args.getString(PROJECT_ID_KEY);
            Assert.assertTrue(!TextUtils.isEmpty(mProjectId));
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_friend_list, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        RelativeLayout friendListLayout = (RelativeLayout) getView();
        Assert.assertTrue(friendListLayout != null);

        mFriendListProgress = friendListLayout.findViewById(R.id.friend_list_progress);
        Assert.assertTrue(mFriendListProgress != null);

        mFriendListRecycler = friendListLayout.findViewById(R.id.friend_list_recycler);
        Assert.assertTrue(mFriendListRecycler != null);

        mFriendListRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));

        mEmptyText = friendListLayout.findViewById(R.id.empty_text);
        Assert.assertTrue(mEmptyText != null);

        if (savedInstanceState != null && savedInstanceState.containsKey(SAVE_STATE_KEY)) {
            mSaveState = savedInstanceState.getParcelable(SAVE_STATE_KEY);
            Assert.assertTrue(mSaveState != null);
        }
    }

    private void initializeFriendPickerFragment(@NonNull FriendPickerFragment friendPickerFragment) {
        Assert.assertTrue(mData != null);

        Set<String> userIds = Stream.of(mFriendListAdapter.mUserDataWrappers)
                .map(userDataWrapper -> userDataWrapper.mUserListData.getId())
                .collect(Collectors.toSet());

        List<FriendPickerFragment.FriendData> friendDatas = Stream.of(mData.getFriendDatas().values())
                .filterNot(friendData -> userIds.contains(friendData.getId()))
                .map(friendData -> new FriendPickerFragment.FriendData(friendData.getId(), friendData.getName(), friendData.getEmail()))
                .collect(Collectors.toList());

        friendPickerFragment.initialize(friendDatas, friendId -> {
            Assert.assertTrue(mData.getFriendDatas().containsKey(friendId));
            Assert.assertTrue(Stream.of(mFriendListAdapter.mUserDataWrappers)
                    .noneMatch(userDataWrapper -> userDataWrapper.mUserListData.getId().equals(friendId)));

            ShowProjectLoader.UserListData friendData = mData.getFriendDatas().get(friendId);
            Assert.assertTrue(friendData != null);

            int position = mFriendListAdapter.getItemCount();

            mFriendListAdapter.mUserDataWrappers.add(new UserDataWrapper(friendData, new HashSet<>()));
            mFriendListAdapter.notifyItemChanged(position);

            if (mData.getUserListDatas().isEmpty()) {
                mEmptyText.setVisibility(View.GONE);
                mFriendListRecycler.setVisibility(View.VISIBLE);
            }
        });
    }

    public void initialize(@Nullable String projectId, @NonNull ShowProjectLoader.Data data) {
        mProjectId = projectId;
        mData = data;

        if (mFriendListAdapter != null)
            mSaveState = mFriendListAdapter.getSaveState();
        else if (mSaveState == null)
            mSaveState = new SaveState(new HashSet<>(), new HashSet<>(), new HashSet<>());

        mFriendListAdapter = new FriendListAdapter(data.getUserListDatas(), mSaveState);
        mFriendListRecycler.setAdapter(mFriendListAdapter);

        mSelectionCallback.setSelected(mFriendListAdapter.getSelected().size());

        mFriendListProgress.setVisibility(View.GONE);

        updateFabVisibility();

        if (mFriendListAdapter.mUserDataWrappers.isEmpty()) {
            mFriendListRecycler.setVisibility(View.GONE);
            mEmptyText.setVisibility(View.VISIBLE);
            mEmptyText.setText(R.string.friends_empty);
        } else {
            mFriendListRecycler.setVisibility(View.VISIBLE);
            mEmptyText.setVisibility(View.GONE);
        }

        FriendPickerFragment friendPickerFragment = (FriendPickerFragment) getChildFragmentManager().findFragmentByTag(FRIEND_PICKER_TAG);
        if (friendPickerFragment != null)
            initializeFriendPickerFragment(friendPickerFragment);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (mFriendListAdapter != null)
            outState.putParcelable(SAVE_STATE_KEY, mFriendListAdapter.getSaveState());
    }

    @SuppressWarnings("RedundantIfStatement")
    public boolean dataChanged() {
        if (mData == null)
            return false;

        SaveState saveState = mFriendListAdapter.getSaveState();

        if (!saveState.mAddedIds.isEmpty())
            return true;

        if (!saveState.mRemovedIds.isEmpty())
            return true;

        return false;
    }

    public void save(@NonNull String name) {
        Assert.assertTrue(!TextUtils.isEmpty(name));
        Assert.assertTrue(mData != null);
        Assert.assertTrue(mFriendListAdapter != null);

        SaveState saveState = mFriendListAdapter.getSaveState();

        if (TextUtils.isEmpty(mProjectId)) {
            Assert.assertTrue(saveState.mRemovedIds.isEmpty());

            DomainFactory.getDomainFactory().createProject(getActivity(), mData.getDataId(), SaveService.Source.GUI, name, saveState.mAddedIds);
        } else {
            DomainFactory.getDomainFactory().updateProject(getActivity(), mData.getDataId(), SaveService.Source.GUI, mProjectId, name, saveState.mAddedIds, saveState.mRemovedIds);
        }
    }

    @Override
    public void setFab(@NonNull FloatingActionButton floatingActionButton) {
        Assert.assertTrue(mFriendListFab == null);

        mFriendListFab = floatingActionButton;

        mFriendListFab.setOnClickListener(v -> {
            FriendPickerFragment friendPickerFragment = FriendPickerFragment.newInstance();
            initializeFriendPickerFragment(friendPickerFragment);
            friendPickerFragment.show(getChildFragmentManager(), FRIEND_PICKER_TAG);
        });

        updateFabVisibility();
    }

    private void updateFabVisibility() {
        if (mFriendListFab == null)
            return;

        if (mData != null && !mSelectionCallback.hasActionMode()) {
            mFriendListFab.show();
        } else {
            mFriendListFab.hide();
        }
    }

    @Override
    public void clearFab() {
        Assert.assertTrue(mFriendListFab != null);

        mFriendListFab.setOnClickListener(null);

        mFriendListFab = null;
    }

    class FriendListAdapter extends RecyclerView.Adapter<FriendListAdapter.FriendHolder> {
        @NonNull
        private final List<UserDataWrapper> mUserDataWrappers;

        FriendListAdapter(@NonNull Collection<ShowProjectLoader.UserListData> userListDatas, @NonNull SaveState saveState) {
            Assert.assertTrue(mData != null);

            Map<String, ShowProjectLoader.UserListData> userListMap = Stream.of(userListDatas)
                    .collect(Collectors.toMap(userListData -> userListData.getId(), userListData -> userListData));

            Stream.of(saveState.mRemovedIds)
                    .forEach(userListMap::remove);

            if (!saveState.mAddedIds.isEmpty()) {
                userListMap.putAll(Stream.of(mData.getFriendDatas().values())
                        .filter(friendData -> saveState.mAddedIds.contains(friendData.getId()))
                        .collect(Collectors.toMap(userListData -> userListData.getId(), userListData -> userListData)));
            }

            mUserDataWrappers = Stream.of(userListMap.values())
                    .sorted((lhs, rhs) -> lhs.getId().compareTo(rhs.getId()))
                    .map(userListData -> new UserDataWrapper(userListData, saveState.mSelectedIds))
                    .collect(Collectors.toList());
        }

        @Override
        public int getItemCount() {
            return mUserDataWrappers.size();
        }

        void unselect() {
            Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .forEach(userDataWrapper -> {
                        userDataWrapper.mSelected = false;
                        notifyItemChanged(mUserDataWrappers.indexOf(userDataWrapper));
                    });
        }

        @Override
        public FriendHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            View friendRow = layoutInflater.inflate(R.layout.row_friend, parent, false);

            TextView friendName = friendRow.findViewById(R.id.friend_name);
            Assert.assertTrue(friendName != null);

            TextView friendEmail = friendRow.findViewById(R.id.friend_email);
            Assert.assertTrue(friendEmail != null);

            return new FriendHolder(friendRow, friendName, friendEmail);
        }

        @Override
        public void onBindViewHolder(final FriendHolder friendHolder, int position) {
            UserDataWrapper userDataWrapper = mUserDataWrappers.get(position);
            Assert.assertTrue(userDataWrapper != null);

            friendHolder.mFriendName.setText(userDataWrapper.mUserListData.getName());
            friendHolder.mFriendEmail.setText(userDataWrapper.mUserListData.getEmail());

            if (userDataWrapper.mSelected)
                friendHolder.mFriendRow.setBackgroundColor(ContextCompat.getColor(getActivity(), R.color.selected));
            else
                friendHolder.mFriendRow.setBackgroundColor(Color.TRANSPARENT);

            friendHolder.mFriendRow.setOnLongClickListener(v -> {
                friendHolder.onLongClick();
                return true;
            });

            friendHolder.mFriendRow.setOnClickListener(v -> {
                if (mSelectionCallback.hasActionMode())
                    friendHolder.onLongClick();
                else
                    friendHolder.onRowClick();
            });
        }

        ArrayList<String> getSelected() {
            return new ArrayList<>(Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .map(userDataWrapper -> userDataWrapper.mUserListData.getEmail())
                    .collect(Collectors.toList()));
        }

        void removeSelected() {
            List<UserDataWrapper> selectedUserDataWrappers = Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .collect(Collectors.toList());

            for (UserDataWrapper userDataWrapper : selectedUserDataWrappers) {
                int position = mUserDataWrappers.indexOf(userDataWrapper);
                mUserDataWrappers.remove(position);
                notifyItemRemoved(position);
            }
        }

        @NonNull
        SaveState getSaveState() {
            Assert.assertTrue(mData != null);

            Set<String> oldUserIds = Stream.of(mData.getUserListDatas())
                    .map(userListData -> userListData.getId())
                    .collect(Collectors.toSet());

            Set<String> newUserIds = Stream.of(mUserDataWrappers)
                    .map(userDataWrapper -> userDataWrapper.mUserListData.getId())
                    .collect(Collectors.toSet());

            Set<String> addedIds = Sets.difference(newUserIds, oldUserIds);
            Set<String> removedIds = Sets.difference(oldUserIds, newUserIds);

            Set<String> selectedIds = Stream.of(mUserDataWrappers)
                    .filter(userDataWrapper -> userDataWrapper.mSelected)
                    .map(userDataWrapper -> userDataWrapper.mUserListData.getId())
                    .collect(Collectors.toSet());

            return new SaveState(addedIds, removedIds, selectedIds);
        }

        class FriendHolder extends RecyclerView.ViewHolder {
            final View mFriendRow;
            final TextView mFriendName;
            final TextView mFriendEmail;

            FriendHolder(@NonNull View friendRow, @NonNull TextView friendName, @NonNull TextView friendEmail) {
                super(friendRow);

                mFriendRow = friendRow;
                mFriendName = friendName;
                mFriendEmail = friendEmail;
            }

            void onRowClick() {
                UserDataWrapper userDataWrapper = mUserDataWrappers.get(getAdapterPosition());
                Assert.assertTrue(userDataWrapper != null);
            }

            void onLongClick() {
                int position = getAdapterPosition();

                UserDataWrapper userDataWrapper = mUserDataWrappers.get(position);
                Assert.assertTrue(userDataWrapper != null);

                userDataWrapper.toggleSelect();
            }
        }
    }

    private class UserDataWrapper {
        @NonNull
        final ShowProjectLoader.UserListData mUserListData;

        boolean mSelected = false;

        UserDataWrapper(@NonNull ShowProjectLoader.UserListData userListData, @NonNull Set<String> selectedIds) {
            mUserListData = userListData;

            mSelected = selectedIds.contains(mUserListData.getId());
        }

        void toggleSelect() {
            mSelected = !mSelected;

            if (mSelected) {
                mSelectionCallback.incrementSelected();
            } else {
                mSelectionCallback.decrementSelected();
            }

            int position = mFriendListAdapter.mUserDataWrappers.indexOf(this);
            Assert.assertTrue(position >= 0);

            mFriendListAdapter.notifyItemChanged(position);
        }
    }

    private static class SaveState implements Parcelable {
        @NonNull
        final Set<String> mAddedIds;

        @NonNull
        final Set<String> mRemovedIds;

        @NonNull
        final Set<String> mSelectedIds;

        SaveState(@NonNull Set<String> addedIds, @NonNull Set<String> removedIds, @NonNull Set<String> selectedIds) {
            mAddedIds = addedIds;
            mRemovedIds = removedIds;
            mSelectedIds = selectedIds;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeStringList(new ArrayList<>(mAddedIds));
            dest.writeStringList(new ArrayList<>(mRemovedIds));
            dest.writeStringList(new ArrayList<>(mSelectedIds));
        }

        @Override
        public int describeContents() {
            return 0;
        }

        public static final Creator<SaveState> CREATOR = new Creator<SaveState>() {
            @Override
            public SaveState createFromParcel(Parcel in) {
                List<String> addedIds = in.createStringArrayList();
                Assert.assertTrue(addedIds != null);

                List<String> removedIds = in.createStringArrayList();
                Assert.assertTrue(removedIds != null);

                List<String> selectedIds = in.createStringArrayList();
                Assert.assertTrue(selectedIds != null);

                return new SaveState(new HashSet<>(addedIds), new HashSet<>(removedIds), new HashSet<>(selectedIds));
            }

            @Override
            public SaveState[] newArray(int size) {
                return new SaveState[size];
            }
        };
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/friends/UserListFragment.kt
package com.krystianwsul.checkme.gui.friends


import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CheckResult
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.CustomItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentFriendListBinding
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.domainmodel.DomainListenerManager
import com.krystianwsul.checkme.domainmodel.extensions.createProject
import com.krystianwsul.checkme.domainmodel.extensions.updateProject
import com.krystianwsul.checkme.gui.base.AbstractFragment
import com.krystianwsul.checkme.gui.base.SnackbarListener
import com.krystianwsul.checkme.gui.main.FabUser
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.tree.AbstractModelNode
import com.krystianwsul.checkme.gui.tree.BaseAdapter
import com.krystianwsul.checkme.gui.tree.HolderType
import com.krystianwsul.checkme.gui.tree.delegates.avatar.AvatarDelegate
import com.krystianwsul.checkme.gui.tree.delegates.avatar.AvatarModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineNameData
import com.krystianwsul.checkme.gui.utils.ResettableProperty
import com.krystianwsul.checkme.gui.utils.SelectionCallback
import com.krystianwsul.checkme.gui.widgets.MyBottomBar
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.utils.tryGetFragment
import com.krystianwsul.checkme.viewmodels.ShowProjectViewModel
import com.krystianwsul.common.utils.ProjectKey
import com.krystianwsul.common.utils.UserKey
import com.krystianwsul.treeadapter.*
import io.reactivex.rxjava3.core.Completable
import kotlinx.parcelize.Parcelize
import java.util.*

class UserListFragment : AbstractFragment(), FabUser {

    companion object {

        private const val SAVE_STATE_KEY = "saveState"
        private const val FRIEND_PICKER_TAG = "friendPicker"

        fun newInstance() = UserListFragment()
    }

    private var projectId: ProjectKey.Shared? = null

    lateinit var treeViewAdapter: TreeViewAdapter<AbstractHolder>
        private set

    private var data: ShowProjectViewModel.Data? = null

    private var saveState = SaveState(HashSet(), HashSet(), HashSet())

    private val selectionCallback = object : SelectionCallback() {

        override val activity get() = requireActivity()

        override fun getTreeViewAdapter() = treeViewAdapter

        override fun unselect(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.unselect(placeholder)

        override val bottomBarData by lazy {
            Triple(listener.getBottomBar(), R.menu.menu_friends, listener::initBottomBar)
        }

        override fun onMenuClick(itemId: Int, placeholder: TreeViewAdapter.Placeholder): Boolean {
            check(itemId == R.id.action_friends_delete)

            (treeViewAdapter.treeModelAdapter as FriendListAdapter).removeSelected(placeholder)

            updateSelectAll()

            return true
        }

        override fun onFirstAdded(placeholder: TreeViewAdapter.Placeholder, initial: Boolean) {
            (activity as AppCompatActivity).startSupportActionMode(this)

            updateFabVisibility()

            super.onFirstAdded(placeholder, initial)
        }

        override fun onLastRemoved(placeholder: TreeViewAdapter.Placeholder) = updateFabVisibility()
    }

    private var friendListFab: FloatingActionButton? = null

    private val listener get() = activity as UserListListener

    private val bindingProperty = ResettableProperty<FragmentFriendListBinding>()
    private var binding by bindingProperty

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) = FragmentFriendListBinding.inflate(inflater, container, false).also { binding = it }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.friendListEmptyTextInclude
                .emptyTextPadding
                .visibility = View.VISIBLE
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        binding.friendListRecycler.layoutManager = LinearLayoutManager(activity)

        if (savedInstanceState?.containsKey(SAVE_STATE_KEY) == true)
            saveState = savedInstanceState.getParcelable(SAVE_STATE_KEY)!!

        initialize()
    }

    private fun initializeFriendPickerFragment(friendPickerFragment: FriendPickerFragment) {
        check(data != null)

        val userIds = getSelected()
                .map { it.userListData.id }
                .toSet()

        val friendDatas = data!!.friendDatas
                .values
                .asSequence()
                .filterNot { userIds.contains(it.id) }
                .map { FriendPickerFragment.FriendData(it.id, it.name, it.email, it.photoUrl) }
                .toList()

        friendPickerFragment.initialize(FriendPickerFragment.Data(data!!.immediate, friendDatas)) { friendId ->
            check(data!!.friendDatas.containsKey(friendId))
            check(getSelected().none { it.userListData.id == friendId })

            val friendData = data!!.friendDatas.getValue(friendId)

            treeViewAdapter.updateDisplayedNodes {
                val userNode = UserNode(friendData, HashSet())

                (treeViewAdapter.treeModelAdapter as FriendListAdapter).apply {
                    userNodes.add(userNode)
                    treeNodeCollection.add(userNode.initialize(treeNodeCollection), it)
                }
            }

            updateVisibility(false)
        }
    }

    fun initialize(projectId: ProjectKey.Shared?, data: ShowProjectViewModel.Data) {
        this.projectId = projectId
        this.data = data

        initialize()
    }

    private fun initialize() {
        if (data == null) return
        if (!bindingProperty.isSet) return

        if (this::treeViewAdapter.isInitialized) {
            saveState = (treeViewAdapter.treeModelAdapter as FriendListAdapter).getSaveState()

            treeViewAdapter.updateDisplayedNodes {
                (treeViewAdapter.treeModelAdapter as FriendListAdapter).initialize(data!!.userListDatas, saveState)

                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, false)
            }
        } else {
            val friendListAdapter = FriendListAdapter()
            friendListAdapter.initialize(data!!.userListDatas, saveState)
            treeViewAdapter = friendListAdapter.treeViewAdapter

            binding.friendListRecycler.apply {
                adapter = treeViewAdapter
                itemAnimator = CustomItemAnimator()
            }

            treeViewAdapter.updateDisplayedNodes {
                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, true)
            }
        }

        updateFabVisibility()
        updateVisibility(data!!.immediate)

        tryGetFragment<FriendPickerFragment>(FRIEND_PICKER_TAG)?.let { initializeFriendPickerFragment(it) }

        updateSelectAll()
    }

    private fun updateVisibility(immediate: Boolean) {
        val hide = mutableListOf<View>(binding.friendListProgress)
        val show: View
        if ((treeViewAdapter.treeModelAdapter as FriendListAdapter).userNodes.isEmpty()) {
            binding.friendListEmptyTextInclude
                    .emptyText
                    .setText(R.string.friends_empty)

            show = binding.friendListEmptyTextInclude.emptyTextLayout
            hide += binding.friendListRecycler
        } else {
            show = binding.friendListRecycler
            hide += binding.friendListEmptyTextInclude.emptyTextLayout
        }

        animateVisibility(listOf(show), hide, immediate)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        if (this::treeViewAdapter.isInitialized) {
            (treeViewAdapter.treeModelAdapter as FriendListAdapter).let {
                outState.putParcelable(SAVE_STATE_KEY, it.getSaveState())
            }
        }
    }

    fun dataChanged(): Boolean {
        if (data == null) return false

        val saveState = (treeViewAdapter.treeModelAdapter as FriendListAdapter).getSaveState()

        if (saveState.addedIds.isNotEmpty()) return true

        return saveState.removedIds.isNotEmpty()
    }

    @CheckResult
    fun save(name: String): Completable {
        check(name.isNotEmpty())
        checkNotNull(data)

        val saveState = (treeViewAdapter.treeModelAdapter as FriendListAdapter).getSaveState()

        return DomainFactory.instance
                .run {
                    if (projectId == null) {
                        check(saveState.removedIds.isEmpty())

                        createProject(
                                DomainListenerManager.NotificationType.All,
                                name,
                                saveState.addedIds,
                        ).ignoreElement()
                    } else {
                        updateProject(
                                DomainListenerManager.NotificationType.All,
                                projectId!!,
                                name,
                                saveState.addedIds,
                                saveState.removedIds,
                        )
                    }
                }
    }

    override fun setFab(floatingActionButton: FloatingActionButton) {
        check(friendListFab == null)

        friendListFab = floatingActionButton

        floatingActionButton.setOnClickListener {
            FriendPickerFragment.newInstance().also {
                initializeFriendPickerFragment(it)
                it.show(childFragmentManager, FRIEND_PICKER_TAG)
            }
        }

        updateFabVisibility()
    }

    private fun updateFabVisibility() {
        friendListFab?.let {
            if (data != null && !selectionCallback.hasActionMode) it.show() else it.hide()
        }
    }

    override fun clearFab() {
        checkNotNull(friendListFab)

        friendListFab = null
    }

    private fun getSelected() = treeViewAdapter.selectedNodes.map { (it.modelNode as UserNode) }

    private fun updateSelectAll() = listener.setUserSelectAllVisibility(treeViewAdapter.displayedNodes.isNotEmpty())

    override fun onDestroyView() {
        bindingProperty.reset()

        super.onDestroyView()
    }

    inner class FriendListAdapter : BaseAdapter(), ActionModeCallback by selectionCallback {

        lateinit var userNodes: MutableList<UserNode>
            private set

        val treeViewAdapter = TreeViewAdapter(
                this,
                TreeViewAdapter.PaddingData(R.layout.row_group_list_fab_padding, R.id.paddingProgress),
        )

        public override lateinit var treeNodeCollection: TreeNodeCollection<AbstractHolder>
            private set

        fun initialize(userListDatas: Collection<ShowProjectViewModel.UserListData>, saveState: SaveState) {
            checkNotNull(data)

            val userListMap = userListDatas.associateBy { it.id }.toMutableMap()

            saveState.removedIds.forEach { userListMap.remove(it) }

            if (saveState.addedIds.isNotEmpty()) {
                userListMap.putAll(data!!.friendDatas.values
                        .filter { saveState.addedIds.contains(it.id) }
                        .associateBy { it.id })
            }

            userNodes = userListMap.values
                    .sortedBy { it.id }
                    .map { UserNode(it, saveState.selectedIds) }
                    .toMutableList()

            treeNodeCollection = TreeNodeCollection(treeViewAdapter)
            treeViewAdapter.setTreeNodeCollection(treeNodeCollection)

            treeNodeCollection.nodes = userNodes.map { it.initialize(treeNodeCollection) }
        }

        fun getSaveState(): SaveState {
            checkNotNull(data)

            val oldUserIds = data!!.userListDatas
                    .map { it.id }
                    .toSet()

            val newUserIds = userNodes.map { it.userListData.id }.toSet()

            val addedIds = newUserIds.minus(oldUserIds)
            val removedIds = oldUserIds.minus(newUserIds)

            val selectedIds = getSelected().map { it.userListData.id }.toSet()

            return SaveState(addedIds, removedIds, selectedIds)
        }

        fun removeSelected(@Suppress("UNUSED_PARAMETER") x: TreeViewAdapter.Placeholder) {
            val selectedUserDataWrappers = getSelected()

            for (userDataWrapper in selectedUserDataWrappers) {
                userNodes.remove(userDataWrapper)
                (treeViewAdapter.treeModelAdapter as FriendListAdapter).treeNodeCollection.remove(userDataWrapper.treeNode, x)
            }
        }
    }

    inner class UserNode(
            val userListData: ShowProjectViewModel.UserListData,
            private val selectedIds: Set<UserKey>,
    ) : AbstractModelNode(), AvatarModelNode, MultiLineModelNode {

        override val holderType = HolderType.AVATAR

        override val name = MultiLineNameData.Visible(userListData.name)

        override val details = Pair(userListData.email, R.color.textSecondary)

        public override lateinit var treeNode: TreeNode<AbstractHolder>
            private set

        override val id = userListData.id

        override val isSelectable = true

        override val parentNode: ModelNode<AbstractHolder>? = null

        override val avatarUrl = userListData.photoUrl

        override val delegates by lazy {
            listOf(
                    AvatarDelegate(this),
                    MultiLineDelegate(this)
            )
        }

        override val widthKey
            get() = MultiLineDelegate.WidthKey(
                    0,
                    true,
                    false,
                    false
            )

        override fun compareTo(other: ModelNode<AbstractHolder>) = userListData.id.compareTo((other as UserNode).userListData.id)

        fun initialize(treeNodeCollection: TreeNodeCollection<AbstractHolder>) = TreeNode(
                this,
                treeNodeCollection,
                selectedIds.contains(userListData.id),
        ).also {
            treeNode = it
            it.setChildTreeNodes(listOf())
        }
    }

    @Parcelize
    class SaveState(
            val addedIds: Set<UserKey>,
            val removedIds: Set<UserKey>,
            val selectedIds: Set<UserKey>,
    ) : Parcelable

    interface UserListListener : SnackbarListener {

        fun setUserSelectAllVisibility(selectAllVisible: Boolean)

        fun getBottomBar(): MyBottomBar

        fun initBottomBar()
    }
}
